rfcs/text/0408-versioning-types.md

```diff
@@ -0,0 +1,344 @@
+      "$ref": "https://blockprotocol.org/@alice/types/property-type/blurb/v/4"
+    }
+  },
+  "required": ["https://blockprotocol.org/@alice/types/property-type/name"]
+}
+```
+
+```json
+{
+  "https://blockprotocol.org/@alice/types/property-type/name": "The Time Machine",
+  "https://blockprotocol.org/@alice/types/property-type/published-on": "1895-05",
+  "https://blockprotocol.org/@alice/types/property-type/blurb": ...
+}
+```
+
+## Determining Type Compatibility
+
+A key part of the reasoning for picking this approach depends on the ability to determine compatibility between various types.
+
+For the purposes of this section, the following assumptions are applied:
+
+- Any given Schema X is _compatible with_ another schema Y if all possible values that satisfy Schema X also satisfy schema Y (note the directionality)
+- Any given Schema X is _equivalent to_ another Schema Y iff X is _compatible with_ Y _and_ Y is _compatible with_ X
+- All Type schemas have `additionalProperties: false` unless otherwise stated
+- 'Constraints' refer to all JSON schema keywords which affect the validation of data in the Block Protocol, these include (but are not limited to) `type`, `properties`, `minItems`, `maxItems`, `minimum`, `maximum`
+- 'Semantic Annotations' refer to all JSON schema keywords which do not affect the validation of data in the Block Protocol, these include (but are not limited to) `title`, `description`, `examples`, `default`
+- The ordering of constraints does not affect the compatibility of schemas
+
+### Data Types
+
+At the present Data Types represent completely disjoint value spaces, and the base primitive types are all **incompatible** with one another. As the [Non-Primitive Data Types RFC](https://github.com/blockprotocol/blockprotocol/pull/355) continues to be specified, it should comment on how checking compatibility will be affected.
+
+### Property Types
+
+1.  Due to how Property Types are used, they implicitly affect the structure of the data in that they require the existence of a key that matches their URI.
+1.  Property Types define their constraints through a `oneOf` field. This `oneOf` is a collection of a number of values (greater than 1) where each value is a variant of the "Property Values" (defined below).
+
+A Property Type `A` is therefore compatible with another Property Type `B` if and only if
+
+1.  They have the same base URI (which implies they are the same property type but perhaps different versions) and,
+1.  Every element of `A`'s `oneOf` is compatible with one element of `B`'s `oneOf`
+
+The "Property Values" is a recursive definition which refers to the sub-schemas defined within the Property Types meta-schema. It defines the element as being one of the following:
+
+#### Data Type Reference
+
+A Data Type reference is an inline reference to a data-type. A Data Type reference will be compatible with another Data Type reference if the Data Types that are referred to are also compatible.
```

Comment:
It's a good call out, I'm not sure how to phrase it more clearly. A Property type is referenced inside a `properties` blob with an associated key. It's necessarily _inside_ some structure inside the schema:

```json
{
  "kind": "entityType",
  "properties": {
    "http://.../property-type/foo": { "ref": "http://.../property-type/foo/v/1" }
  }
  ...
}
```

Whereas data-types are "$ref" inline so they appear at the top-level of the schema:

```json
{
  "kind": "dataType",
  "$ref": "http://.../data-type/bar/v/1",
  ...
}
```

Reply:
Still not following this, might be worth a sync discussion - surely a data type is referred to from inside a propertyType schema?

```json
{
  "kind": "propertyType",
  "$id": "http://.../property-type/foo/v/1",
  "oneOf: [ {   "$ref": "http://.../data-type/bar/v/1" } ]
}
```
where the data type looks like this:
```
{
  "kind": "dataType",
  "$id": "http://.../data-type/bar/v/1",
  "type": "string"
}

Reply:
Yes, so the example you gave has the data-type's constraints being applied at the same level as `"kind": "propertyType"` (because "oneOf" basically collapses it in)

If you take this entityType:

```json
{
  "kind": "entityType",
  "$id": "http://.../entity-type/baz/v/1",
  "type": "object",
  "properties": {
    "http://.../property-type/foo/v/1": { "$ref": "http://.../property-type/foo/v/1" }
  }
}
```

Then "http://.../property-type/foo/v/1" does _not_ apply at the top level of the entityType object, its constraints apply to the "http://.../property-type/foo/v/1" _property_ within that object.

Thus the _constraints_ on a property type being the same as the constraints on another property type isn't enough to confirm compatibility because there's an implicit effect on the data (which we have added) which says the property name has to be equal to the `base_uri`.

Reply:
Right, understood – but the `$ref` isn't inline in my head – my confusion is thinking that there's no scenario in which this earlier example appears:
```json
{
  "kind": "dataType",
  "$ref": "http://.../data-type/bar/v/1",
  ...
}
```
because if we take the property type:
```json
{
  "kind": "propertyType",
  "$id": "http://.../property-type/foo/v/1",
  "oneOf": [ { "$ref": "http://.../data-type/bar/v/1" } ]
}
```
if we evaluate it, isn't the `$ref` is replaced by the schema it refers to, so you never have the `$ref` appear alongside `kind: "dataType"`? I may be misunderstanding how evaluation works and it's (the `$ref`) left lying around.

Reply:
I don't think this is important to merging the RFC, but might be revealing some misunderstanding I have that's worth clearing up for later.

Reply:
I don't think it's confusion on your end, I think it's me misusing words because I'm lacking a better way of describing it.

- The `$ref` _is_ replaced
- The `oneOf` means it's not 'inline'

I'm just trying to say the constraints within the data-type end up getting applied (if there's only one option in the `oneOf` I guess) as if they had been written inline the data-type object at the top level. That doesn't apply to property types, where there is some more structure implicitly captured that we need to think about for compatibility.

Reply:
I've removed the mention of "inline" because I don't think it adds value to the sentence and is just causing confusion.

Reply:
Understood, seems reasonable